consensus/hotstuff/verification/combined_verifier_v3.go

```diff
@@ -119,60 +139,84 @@ func (c *CombinedVerifierV3) VerifyQC(signers flow.IdentityList, sigData []byte,
 	}
 
 	// verify the aggregated staking and beacon signatures next (more costly)
-
+	// Caution: this function will error if pubKeys is empty
 	verifyAggregatedSignature := func(pubKeys []crypto.PublicKey, aggregatedSig crypto.Signature, hasher hash.Hasher) error {
-		// TODO: to be replaced by model/signature.PublicKeyAggregator in V2
-		aggregatedKey, err := crypto.AggregateBLSPublicKeys(pubKeys)
+		// TODO: as further optimization, replace the following call with model/signature.PublicKeyAggregator
+		aggregatedKey, err := crypto.AggregateBLSPublicKeys(pubKeys) // caution: requires non-empty slice of keys!
 		if err != nil {
 			return fmt.Errorf("internal error computing aggregated key: %w", err)
 		}
 		valid, err := aggregatedKey.Verify(aggregatedSig, msg, hasher)
 		if err != nil {
 			return fmt.Errorf("internal error while verifying aggregated signature: %w", err)
 		}
-
 		if !valid {
-			return fmt.Errorf("invalid aggregated staking sig for block %v: %w", block.BlockID, model.ErrInvalidSignature)
+			return fmt.Errorf("invalid aggregated sig for block %v: %w", block.BlockID, model.ErrInvalidSignature)
 		}
-
 		return nil
 	}
 
-	// first fetch all beacon signers public keys
-	beaconPubKeys := make([]crypto.PublicKey, 0, len(blockSigData.RandomBeaconSigners))
+	// STEP 2: verify aggregated random beacon key shares
+	// Step 2a: fetch all beacon signers public keys.
+	// Note: A valid random beacon group sig is required for QC validity. To reconstruct
+	// the group sig, _strictly more_ than `threshold` sig shares are required.
+	threshold := msig.RandomBeaconThreshold(int(dkg.Size()))
+	numRbSigners := len(blockSigData.RandomBeaconSigners)
+	if numRbSigners <= threshold {
+		return fmt.Errorf("require at least %d random beacon sig shares but only got %d: %w", threshold+1, numRbSigners, model.ErrInvalidFormat)
```

Comment:
This check is to mitigate grieving attacks, i.e. depriving honest participants of their earned rewards. Some more context:
* Participating in the DKG is extra work. We encourage consensus participants to sign with their random beacon key, so that they contribute to HotStuff as well as the random beacon. While we allow consensus participants to only sign  with their staking key, it is discouraged because then, the participant does not contribute to the random beacon. 
* Likely, we will "encourage" nodes via their rewards: if they sign with their random beacon keys, they get a bit more rewards vs. only providing a staking sig. This also compensates the nodes for their extra effort participating in the DKG as opposed to being lazy and skipping the DKG. 

Lets consider the case where we see a QC:
1. a supermajority of consensus participants have signed either with their staking or random beacon key
2. the QC has a valid random beacon group signature 
3. _but_ the random beacon signers are insufficient 

At this point, we know something fishy is going on:
* The node that constructed the QC must have collected enough random beacon key shares. Otherwise, it would not be able to reconstruct the group signature. 
* BUT, the node is not giving credit to some of these random beacon signers for their contribution. Why not?
   * In case the network gives extra reward to random beacon signers, this is clearly a grieving attack. 
   * But even if we don't pay extra rewards, an honest node is supposed to credit the random beacon signers in its QC. If the node reports fewer random beacon signers than the minimally needed number, it is purposefully miss-representing node contributions. Not for rewards, but probably for some other off-chain benefits (e.g. attracting more delegators by under-reporting contributions of other nodes).

Reply:
added the following documentation to the code: https://github.com/onflow/flow-go/blob/44d665742ac284822af7233fc564a766a6837d33/consensus/hotstuff/verification/combined_verifier_v3.go#L165-L172